src/Language/PureScript/CodeGen/JS/Optimizer/TCO.hs

```diff
@@ -43,65 +48,76 @@ tco = everywhereOnJS convert where
+      match :: JS -> Int
+      match (JSVar _ ident') | ident == ident' = 1
+      match _ = 0
+
+    allInTailPosition (JSReturn _ expr)
+      | isSelfCall ident expr = countSelfReferences expr == 1
+      | otherwise = countSelfReferences expr == 0
+    allInTailPosition (JSWhile _ js1 body)
+      = countSelfReferences js1 == 0 && allInTailPosition body
+    allInTailPosition (JSFor _ _ js1 js2 body)
+      = countSelfReferences js1 == 0 && countSelfReferences js2 == 0 && allInTailPosition body
+    allInTailPosition (JSForIn _ _ js1 body)
+      = countSelfReferences js1 == 0 && allInTailPosition body
+    allInTailPosition (JSIfElse _ js1 body el)
+      = countSelfReferences js1 == 0 && allInTailPosition body && all allInTailPosition el
+    allInTailPosition (JSBlock _ body)
+      = all allInTailPosition body
+    allInTailPosition (JSLabel _ _ body)
+      = allInTailPosition body
+    allInTailPosition _
+      = False
 
   toLoop :: Text -> [Text] -> JS -> JS
-  toLoop ident allArgs js = JSBlock rootSS $
+  toLoop ident allArgs js =
+      JSBlock rootSS $
         map (\arg -> JSVariableIntroduction rootSS arg (Just (JSVar rootSS (copyVar arg)))) allArgs ++
-        [ JSLabel rootSS tcoLabel $ JSWhile rootSS (JSBooleanLiteral rootSS True) (JSBlock rootSS [ everywhereOnJS loopify js ]) ]
+        [ JSVariableIntroduction rootSS tcoDone (Just (JSBooleanLiteral rootSS False))
+        , JSVariableIntroduction rootSS tcoResult Nothing
+        ] ++
+        map (\arg ->
+          JSVariableIntroduction rootSS (tcoVar arg) Nothing) allArgs ++
+        [ JSFunction rootSS (Just tcoLoop) allArgs (JSBlock rootSS [loopify js])
+        , JSLabel rootSS tcoLabel $
+            JSWhile rootSS (JSUnary rootSS Not (JSVar rootSS tcoDone))
+              (JSBlock rootSS
+                (JSAssignment rootSS (JSVar rootSS tcoResult) (JSApp rootSS (JSVar rootSS tcoLoop) (map (JSVar rootSS) allArgs))
+                : map (\arg ->
+                    JSAssignment rootSS (JSVar rootSS arg) (JSVar rootSS (tcoVar arg))) allArgs))
+        , JSReturn rootSS (JSVar rootSS tcoResult)
+        ]
     where
     rootSS = Nothing
 
     loopify :: JS -> JS
-    loopify (JSReturn ss ret) | isSelfCall ident ret =
-      let
-        allArgumentValues = concat $ collectSelfCallArgs [] ret
-      in
-        JSBlock ss $ zipWith (\val arg ->
-                    JSVariableIntroduction ss (tcoVar arg) (Just val)) allArgumentValues allArgs
-                  ++ map (\arg ->
-                    JSAssignment ss (JSVar ss arg) (JSVar ss (tcoVar arg))) allArgs
-                  ++ [ JSContinue ss tcoLabel ]
+    loopify (JSReturn ss ret)
+      | isSelfCall ident ret =
+        let
+          allArgumentValues = concat $ collectArgs [] ret
+        in
+          JSBlock ss $
+            zipWith (\val arg ->
+              JSAssignment ss (JSVar ss (tcoVar arg)) val) allArgumentValues allArgs
+              ++ [ JSReturn ss (JSVar ss "undefined") ]
```

Comment:
Does the PureScript compiler ever generate code with early returns like that? I figured this would only ever occur at the natural end of execution anyway. But it doesn't hurt I guess. I mostly just didn't like the reference to `undefined`. `return void 0` or `return` or simply leaving it out would all be preferable.